type switch on binary interface

### The following program `sample.go` triggers an unexpected result

```go
package main

import ( 
    "io"
    "os" 
)       
    
func f(i interface{}) {
    switch at := i.(type) {
    case int, int8:
        println("integer", at)
    case io.Reader:
        println("reader")
    }
    println("bye")
}   

func main() {
    var fd *os.File
    var r io.Reader = fd
    f(r)
}
```


### Expected result

```console
$ go run ./sample.go
reader
bye
```


### Got

```console
$ yaegi ./sample.go
run.go:2996: ./sample.go:10:2: case interface {} &{<nil> 17 [] <nil> <nil> <nil> [] [] <nil> [] int  0 <nil> <nil> <nil> int false false false}
/Users/marc/yt/case_nil.go:9:2: panic
/Users/marc/yt/case_nil.go:17:9: panic
run: interface conversion: interface {} is *os.File, not interp.valueInterface
goroutine 1 [running]:
runtime/debug.Stack()
	/Users/marc/sdk/go1.17.5/src/runtime/debug/stack.go:24 +0x88
github.com/traefik/yaegi/interp.(*Interpreter).Execute.func1(0x140000a5ae0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/program.go:131 +0x8c
panic({0x102c60dc0, 0x14000110630})
	/Users/marc/sdk/go1.17.5/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp.runCfg.func1(0x140001e2370, 0x140000a8480, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:194 +0x164
panic({0x102c60dc0, 0x14000110630})
	/Users/marc/sdk/go1.17.5/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp.runCfg.func1(0x140001e2420, 0x1400000c240, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:194 +0x164
panic({0x102c60dc0, 0x14000110630})
	/Users/marc/sdk/go1.17.5/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp._case.func5(0x140001e2420)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:3015 +0x2b4
github.com/traefik/yaegi/interp.runCfg(0x1400000c240, 0x140001e2420, 0x140001daea0, 0x140000a9320)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:202 +0x268
github.com/traefik/yaegi/interp.call.func9(0x140001e2370)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:1433 +0x7c0
github.com/traefik/yaegi/interp.runCfg(0x140000a8480, 0x140001e2370, 0x1400000d8c0, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:202 +0x268
github.com/traefik/yaegi/interp.(*Interpreter).run(0x140003e0000, 0x1400000d8c0, 0x140003e2000)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:121 +0x3d4
github.com/traefik/yaegi/interp.(*Interpreter).Execute(0x140003e0000, 0x14000110420)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/program.go:157 +0x1e0
github.com/traefik/yaegi/interp.(*Interpreter).eval(0x140003e0000, {0x140001fc380, 0xd4}, {0x16dc677a0, 0x1a}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:615 +0x98
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0x140003e0000, {0x16dc677a0, 0x1a})
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:508 +0xe8
main.runFile(0x140003e0000, {0x16dc677a0, 0x1a}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:153 +0xb0
main.run({0x140001be010, 0x1, 0x1})
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:116 +0xe14
main.main()
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x384
```


### Yaegi Version

v0.11.1

### Additional Notes

this is issue is raised when running yaegi on `github.com/mediocregopher/radix/v4`

--- a/_test/issue-1337.go
+++ b/_test/issue-1337.go
@@ -1,0 +1,26 @@
+package main
+
+import (
+	"io"
+	"os"
+)
+
+func f(i interface{}) {
+	switch at := i.(type) {
+	case int, int8:
+		println("integer", at)
+	case io.Reader:
+		println("reader")
+	}
+	println("bye")
+}
+
+func main() {
+	var fd *os.File
+	var r io.Reader = fd
+	f(r)
+}
+
+// Output:
+// reader
+// bye

--- a/interp/run.go
+++ b/interp/run.go
@@ -2956,13 +2956,21 @@ func _case(n *node) {
 						return tnext
 					}
 					rtyp := typ.TypeOf()
-					if rtyp != nil && rtyp.String() == t.String() && implementsInterface(v, typ) {
-						destValue(f).Set(v.Elem())
+					if rtyp == nil {
+						return fnext
+					}
+					elem := v.Elem()
+					if rtyp.String() == t.String() && implementsInterface(v, typ) {
+						destValue(f).Set(elem)
 						return tnext
 					}
 					ival := v.Interface()
-					if ival != nil && rtyp != nil && rtyp.String() == reflect.TypeOf(ival).String() {
-						destValue(f).Set(v.Elem())
+					if ival != nil && rtyp.String() == reflect.TypeOf(ival).String() {
+						destValue(f).Set(elem)
+						return tnext
+					}
+					if typ.cat == valueT && rtyp.Kind() == reflect.Interface && elem.IsValid() && elem.Type().Implements(rtyp) {
+						destValue(f).Set(elem)
 						return tnext
 					}
 					return fnext
@@ -2980,12 +2988,37 @@ func _case(n *node) {
 				}
 				return fnext
 			}
+
 		default:
-			// TODO(mpl): probably needs to be fixed for empty interfaces, like above.
-			// match against multiple types: assign var to interface value
 			n.exec = func(f *frame) bltn {
 				val := srcValue(f)
-				if v := srcValue(f).Interface().(valueInterface).node; v != nil {
+				if t := val.Type(); t.Kind() == reflect.Interface {
+					for _, typ := range types {
+						if typ.cat == nilT && val.IsNil() {
+							return tnext
+						}
+						rtyp := typ.TypeOf()
+						if rtyp == nil {
+							continue
+						}
+						elem := val.Elem()
+						if rtyp.String() == t.String() && implementsInterface(val, typ) {
+							destValue(f).Set(elem)
+							return tnext
+						}
+						ival := val.Interface()
+						if ival != nil && rtyp.String() == reflect.TypeOf(ival).String() {
+							destValue(f).Set(elem)
+							return tnext
+						}
+						if typ.cat == valueT && rtyp.Kind() == reflect.Interface && elem.IsValid() && elem.Type().Implements(rtyp) {
+							destValue(f).Set(elem)
+							return tnext
+						}
+					}
+					return fnext
+				}
+				if v := val.Interface().(valueInterface).node; v != nil {
 					for _, typ := range types {
 						if v.typ.id() == typ.id() {
 							destValue(f).Set(val)

